ledger/metrics/src/main/scala/com/daml/metrics/grpc/GrpcMetricsServerInterceptor.scala

```diff
@@ -0,0 +1,150 @@
+  MetricsGrpcClientType,
+  MetricsGrpcMethodName,
+  MetricsGrpcResponseCode,
+  MetricsGrpcServerType,
+  MetricsGrpcServiceName,
+  MetricsRequestTypeStreaming,
+  MetricsRequestTypeUnary,
+}
+import com.google.protobuf.{GeneratedMessage => JavaGeneratedMessage}
+import io.grpc.ForwardingServerCall.SimpleForwardingServerCall
+import io.grpc.ServerCall.Listener
+import io.grpc.{
+  ForwardingServerCallListener,
+  Metadata,
+  ServerCall,
+  ServerCallHandler,
+  ServerInterceptor,
+  Status,
+}
+import scalapb.{GeneratedMessage => ScalapbGeneratedMessage}
+
+class GrpcMetricsServerInterceptor(metrics: GrpcServerMetrics) extends ServerInterceptor {
+
+  override def interceptCall[ReqT, RespT](
+      serverCall: ServerCall[ReqT, RespT],
+      metadata: Metadata,
+      serverCallHandler: ServerCallHandler[ReqT, RespT],
+  ): ServerCall.Listener[ReqT] = {
+    val methodDescriptor = serverCall.getMethodDescriptor
+    val clientType =
+      if (methodDescriptor.getType.clientSendsOneMessage()) MetricsRequestTypeUnary
+      else MetricsRequestTypeStreaming
+    val serverType =
+      if (methodDescriptor.getType.serverSendsOneMessage()) MetricsRequestTypeUnary
+      else MetricsRequestTypeStreaming
+    withMetricLabels(
+      MetricsGrpcServiceName -> methodDescriptor.getServiceName,
+      MetricsGrpcMethodName -> methodDescriptor.getBareMethodName,
+      MetricsGrpcClientType -> clientType,
+      MetricsGrpcServerType -> serverType,
+    ) { implicit mc =>
+      val timerHandle = metrics.callTimer.startAsync()
+      metrics.callsStarted.mark()
+      val metricsServerCall = new MetricsCall(
+        serverCall,
+        timerHandle,
+        metrics.messagesSent,
+        metrics.messagesSentSize,
+        metrics.callsFinished,
+        metrics.messagesRequested,
+      )
+      new MonitoringServerCallListener(
+        serverCallHandler.startCall(metricsServerCall, metadata),
+        metrics.messagesReceived,
+        metrics.messagesReceivedSize,
+      )
+    }
+  }
+
+  private final class MonitoringServerCallListener[T](
+      val delegate: Listener[T],
+      messagesReceived: Meter,
+      messagesReceivedSize: Histogram,
+  )(implicit
+      metricsContext: MetricsContext
+  ) extends ForwardingServerCallListener[T] {
+
+    override def onMessage(message: T): Unit = {
+      super.onMessage(message)
+      updateHistogramWithSerializedSize(messagesReceivedSize, message)
+      messagesReceived.mark()
+    }
+  }
+
+  private final class MetricsCall[ReqT, RespT](
+      call: ServerCall[ReqT, RespT],
+      timer: TimerHandle,
+      messagesSent: Meter,
+      messagesSentSize: Histogram,
+      callsClosed: Meter,
+      messagesRequested: Meter,
+  )(implicit metricsContext: MetricsContext)
+      extends SimpleForwardingServerCall[ReqT, RespT](call) {
+
+    override def request(numMessages: Int): Unit = {
+      super.request(numMessages)
+      messagesRequested.mark(numMessages.toLong)
```

Comment:
I don't think it makes sense to monitor this value.
It is part of the internal backpressure support of gRPC. It will be matching the number of received messages
I cannot think of a use case when this value would be useful.

Reply:
I was thinking with regards to streaming, it could allow us to see any type of bottlenecks where back-pressure is applied and we see this rate going up and down.

Reply:
I really don't see how.
This number follows exactly the number of sent messages. The only difference is that the number of sent messages grows by 1 each time, and the requested number can grow in steps.
If we had the number of messages ready and waiting to be sent, that could be interesting, but this number doesn't give that.

Reply:
Depending on the implementation these numbers can be decoupled. If the server does not currently have enough messages to send, the request would increase while the sent would be constant until the server is capable of sending the messages.

Reply:
No, the request will not increase if the server is not sending message.
The `request(n)` function is the client telling the server 'you can send me n additional messages until my internal buffer is full'.
For one stream, if `N` is the sum of the values sent using `request(n)`, `b` the size of the client internal buffer and `m` the  number of messages sent:
* `0 <= N - m <=  b`
* `N - m = 0` means backpressure, the server is not allowed to send more messages
* `N - m =  b` means upstream has processed all received messages, its internal buffer is empty.

The difference between `m` and `N` can never be bigger than `b`, otherwise the server would be allowed to send at once more messages that the client says it can handle.

Reply:
Hmm, I do see it now, this metrics is indeed pretty useless. Will remove it